compiler/rustc_passes/src/dead.rs

```diff
@@ -789,7 +843,33 @@ impl<'tcx> Visitor<'tcx> for DeadVisitor<'tcx> {
 
     fn visit_field_def(&mut self, field: &'tcx hir::FieldDef<'tcx>) {
         if self.should_warn_about_field(&field) {
-            self.warn_dead_code(field.hir_id, field.span, field.ident.name, "read");
+            let def_id = self.tcx.hir().local_def_id(field.hir_id);
+            if !field.is_positional() {
+                self.warn_dead_code(field.hir_id, field.span, field.ident.name, "read");
+            } else if !self.tcx.type_of(def_id).is_unit() {
```

Comment:
This doesn't seem like the only type which is expected to be set but not read. For example, PhantomData (and probably other ZSTs, like PhantomPinned?) are typically created but not read, and likely shouldn't be read (it doesn't really make sense to). I'm not sure what class of fields we should consider here, but maybe ZSTs aren't bad. I don't know if we have layout information here though.

Reply:
Hmm, we have some sort of query we use when checking for 1-ZSTs in `repr(transparent)`.  Maybe this lint should ignore anything which that would let through?  It must already handle things like "that's a type variable so I can't assume its layout", which seems like something still worth warning about even if it turns out the parameter is always a ZST.

Reply:
Would the following work in this context?

```
    self.tcx.layout_of(self.tcx.type_of(def_id).map_or(true, |layout| layout.is_zst())
```

Reply:
If this does indeed work, can we add it and make sure that a test covers this (by having it not trigger for `()` and some `struct Void;`?